packages/components/src/internal/components/assay/withAssayModels.tsx

```diff
@@ -0,0 +1,286 @@
+    loadDefinitions: fetchAllAssays,
+    loadProtocol: fetchProtocol,
+};
+
+/**
+ * Provides a wrapped component with assay definitions. These definitions are loaded into the
+ * [[AssayStateModel]] which is injected into the component as the "assayModel" property.
+ * Optionally, if the "assayName" property is specified it will load the associated assay
+ * protocol and pass it, along with it's specific assay definition, as props "assayProtocol"
+ * and "assayDefinition".
+ * @param ComponentToWrap: The component definition (e.g. class, function) to wrap.
+ * This will have [[InjectedAssayModel]] props injected into it when instantiated.
+ * @param defaultProps: Provide alternative "defaultProps" for this wrapped component.
+ */
+export function withAssayModels<Props>(
+    ComponentToWrap: ComponentType<Props & InjectedAssayModel>,
+    defaultProps?: WithAssayModelProps
+): ComponentType<Props & WithAssayModelProps> {
+    type WrappedProps = Props & WithAssayModelProps;
+
+    class ComponentWithAssays extends PureComponent<WrappedProps, State> {
+        static defaultProps;
+
+        state: Readonly<State> = produce({}, () => ({
+            context: { assayDefinition: undefined, assayProtocol: undefined },
+            model: new AssayStateModel(),
+        }));
+
+        private _mounted = false;
+
+        componentDidMount = (): void => {
+            this._mounted = true;
+            this.load();
+        };
+
+        componentDidUpdate = (prevProps: WrappedProps): void => {
+            if (this.props.assayName !== prevProps.assayName) {
+                this.load();
+            }
+        };
+
+        componentWillUnmount = (): void => {
+            this._mounted = false;
+        };
+
+        load = async (): Promise<void> => {
+            await this.loadDefinitions();
+            await this.loadProtocol();
+        };
+
+        loadDefinitions = async (): Promise<void> => {
+            const { assayLoader } = this.props;
+            const { model } = this.state;
+
+            if (model.definitionsLoadingState === LoadingState.LOADED) {
+                return;
+            }
+
+            this.updateModel({ definitionsError: undefined, definitionsLoadingState: LoadingState.LOADING });
+
+            try {
+                const definitions = await assayLoader.loadDefinitions();
+
+                this.updateModel({
+                    definitions: definitions.toArray(),
+                    definitionsLoadingState: LoadingState.LOADED,
+                });
+            } catch (definitionsError) {
+                this.updateModel({ definitionsError, definitionsLoadingState: LoadingState.LOADED });
+            }
+        };
+
+        loadProtocol = async (): Promise<void> => {
+            const { assayLoader, assayName } = this.props;
+            const { model } = this.state;
+
+            if (!assayName && model.protocolLoadingState === LoadingState.INITIALIZED) {
+                return;
```

Comment:
any reason that this should set the `protocolLoadingState` to LOADED here instead of leaving it as INITIALIZED?

Reply:
I think we should only set `LOADED` after we've loaded and leave this check as-is. I've added a comment to make it a little clearer what this check is doing.

```ts
// If an "assayName" is not provided and one has not ever been loaded by this instance,
// then do not attempt to process the "assayName" as it is an optional behavior to load the protocol.
if (!assayName && model.protocolLoadingState === LoadingState.INITIALIZED) {
    return;
}
```